Helm: add support for setting an s3 backoff_config via the loki helm chart

rl0nergan opened the pull request:
**What this PR does / why we need it**:
Currently the Loki helm chart does not support defining s3 backoff config settings. This PR adds support to the loki.commonStorageConfig helper to add backoff config options when they're specified in the helm values
**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/loki?pullRequest=11458) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/loki?pullRequest=11458) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/loki?pullRequest=11458) it.</sub>

rl0nergan commented:
hey @MichelHollands I saw you reviewed a few other helm chart PRs recently so figured I'd ping you. I'll get all the versions bumped again, but can I please get a review on this? Thanks

MichelHollands commented:
@rl0nergan Thanks for this contribution. Please update the documentation as well in addition to bumping the version. The https://grafana.com/docs/loki/latest/setup/install/helm/reference/ page is derived from production/helm/loki/reference.md. It has a `loki.storage` Helm value that needs to have the new section.  After updating it you need to run `make -C docs sources/setup/install/helm/reference.md` to update the docs sources.

MichelHollands commented on production/helm/loki/values.yaml:
```diff
@@ -285,6 +285,13 @@ loki:
       s3ForcePathStyle: false
       insecure: false
       http_config: {}
+      backoff_config:
```
This should be the empty object so existing customers don't get different behaviour when upgrading.

MichelHollands replied:
This isn't empty yet. Are you still working on this?

rl0nergan replied:
whoops! didn't pull after committing your suggestion on the version so missed that it didn't get pushed up. this should be empty now and the docs updated to reflect it

MichelHollands commented on production/helm/loki/templates/_helpers.tpl:
```diff
@@ -245,7 +245,20 @@ s3:
     ca_file: {{ . }}
     {{- end}}
   {{- end }}
+  {{- with .backoff_config}}
```
Can you add an if statement where to check if backoff_config is set? If it's not set/empty then this should not create anything.

rl0nergan replied:
I don't think an if statement is needed here. tested this with a few different values and if the backoff_config isn't specified or is an empty object it just doesn't get added to the generated config:

1. backoff_config is set
    * values:
    ```
      storage:
        bucketNames:
          chunks: loki-chunks
          ruler: loki-ruler
          admin: loki-admin
        type: s3
        s3:
          endpoint: https://some.hostname.com:443
          accessKeyId: ACCESS_KEY
          secretAccessKey: SECRET_KEY
          s3ForcePathStyle: false
          insecure: false
          backoff_config:
            min_period: "2s"
            max_period: "5s"
            max_retries: 5
    ```

    * resulting s3 configuration:
    ```
      storage:
        s3:
          access_key_id: ACCESS_KEY
          backoff_config:
            max_period: 5s
            max_retries: 5
            min_period: 2s
          bucketnames: loki-chunks
          endpoint: https://some.hostname.com:443
          insecure: false
          s3forcepathstyle: false
          secret_access_key: SECRET_KEY
    ```

2. backoff_config with empty object:
    * values:
    ```
      storage:
        bucketNames:
          chunks: loki-chunks
          ruler: loki-ruler
          admin: loki-admin
        type: s3
        s3:
          endpoint: https://some.hostname.com:443
          accessKeyId: ACCESS_KEY
          secretAccessKey: SECRET_KEY
          s3ForcePathStyle: false
          insecure: false
          backoff_config: {}
    ```

    * resulting s3 configuration:
    ```
      storage:
        s3:
          access_key_id: ACCESS_KEY
          bucketnames: loki-chunks
          endpoint: https://some.hostname.com:443
          insecure: false
          s3forcepathstyle: false
          secret_access_key: SECRET_KEY
    ```
2. backoff_config not specified:
    * values:
    ```
      storage:
        bucketNames:
          chunks: loki-chunks
          ruler: loki-ruler
          admin: loki-admin
        type: s3
        s3:
          endpoint: https://some.hostname.com:443
          accessKeyId: ACCESS_KEY
          secretAccessKey: SECRET_KEY
          s3ForcePathStyle: false
          insecure: false
    ```

    * resulting s3 configuration:
    ```
      storage:
        s3:
          access_key_id: ACCESS_KEY
          bucketnames: loki-chunks
          endpoint: https://some.hostname.com:443
          insecure: false
          s3forcepathstyle: false
          secret_access_key: SECRET_KEY
    ```

MichelHollands replied:
You are right. I should have read the docs for the with statement:
```
{{with pipeline}} T1 {{end}}
	If the value of the pipeline is empty, no output is generated;
	otherwise, dot is set to the value of the pipeline and T1 is
	executed.
```

MichelHollands commented on production/helm/loki/README.md:
```diff
@@ -1,6 +1,6 @@
 # loki
 
-![Version: 5.41.2](https://img.shields.io/badge/Version-5.41.2-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 2.9.3](https://img.shields.io/badge/AppVersion-2.9.3-informational?style=flat-square)
+![Version: 5.41.3](https://img.shields.io/badge/Version-5.41.2-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 2.9.3](https://img.shields.io/badge/AppVersion-2.9.3-informational?style=flat-square)
```
```suggestion
![Version: 5.41.3](https://img.shields.io/badge/Version-5.41.3-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 2.9.3](https://img.shields.io/badge/AppVersion-2.9.3-informational?style=flat-square)
```
Missed one version change here.

MichelHollands commented:
@rl0nergan The helm-ci step failed. You need to change the version once more in that line. I've added a suggestion for it.
Also, are the backoff values you provide the default ones? This section should be optional config so existing users don't see a change if they upgrade.

rl0nergan commented:
> @rl0nergan The helm-ci step failed. You need to change the version once more in that line. I've added a suggestion for it. Also, are the backoff values you provide the default ones? This section should be optional config so existing users don't see a change if they upgrade.

Done and addressed comments. The values I had specified were defaults, but I changed it to an empty object and added a comment linking to the Loki documentation for how to set those like is done for other config options instead.

MichelHollands commented:
@rl0nergan One more CI check failed: it looks like just an extra space. I'll merge it once that is fixed.

rl0nergan commented:
@MichelHollands just pushed that up. forgot to run make on the docs after removing the trailing space from the values file yesterday 😅. should hopefully be good to go now

MichelHollands reviewed (approved):
LGTM
